contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,659 @@
+        type ContractError = DexError;
+    }
+
+    impl SimpleDex {
+        #[ink(constructor)]
+        pub fn new(ubik: AccountId, cyberiad: AccountId, lono: AccountId) -> Self {
+            let caller = Self::env().caller();
+            let code_hash = Self::env()
+                .own_code_hash()
+                .expect("Called new on a contract with no code hash");
+            let required_role = Role::Initializer(code_hash);
+            let access_control = AccountId::from(ACCESS_CONTROL_PUBKEY);
+
+            let role_check = <Self as AccessControlled>::check_role(
+                access_control,
+                caller,
+                required_role,
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            );
+
+            match role_check {
+                Ok(_) => {
+                    initialize_contract(|contract| Self::new_init(contract, ubik, cyberiad, lono))
+                }
+                Err(why) => panic!("Could not initialize the contract {:?}", why),
+            }
+        }
+
+        /// Swaps the transferred amount of native token to one of the pools PSP22 tokens
+        #[ink(message, payable)]
+        pub fn native_to_token(&mut self, token_out: AccountId) -> Result<(), DexError> {
+            let amount_token_in = self.env().transferred_value();
+            // we use the balance of the native token just before the exchange tx
+            let balance_token_in = self
+                .env()
+                .balance()
+                .checked_sub(amount_token_in)
+                .ok_or(DexError::Arithmethic)?;
+
+            let swap_fee = self.swap_fee;
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            let balance_token_out = self.balance_of(token_out, this)?;
+
+            let amount_token_out = Self::out_given_in(
+                amount_token_in,
+                balance_token_in,
+                balance_token_out,
+                swap_fee,
+            )?;
+
+            // transfer token_out from contract to user
+            self.transfer_tx(token_out, caller, amount_token_out)??;
+
+            // emit event
+            Self::emit_event(
+                self.env(),
+                Event::NativeToToken(NativeToToken {
+                    caller,
+                    token_out,
+                    amount_in: amount_token_in,
+                    amount_out: amount_token_out,
+                }),
+            );
+
+            Ok(())
+        }
+
+        /// Swap the specified amount of one of the pools PSP22 tokens to the native token
+        /// Calling account needs to give allowance to the DEX contract to spend `amount_token_in` of `token_in` on its behalf
+        /// before executing this tx.
+        #[ink(message)]
+        pub fn token_to_native(
+            &mut self,
+            token_in: AccountId,
+            amount_token_in: u128,
```

Comment:
Normally there should also be a third argument here `min_amount_token_out` that is to deal with slippage. Maybe it is not necessary, but I guess MASSIVE sandwich attacks are possible without that :P